app.py

```diff
@@ -77,9 +78,17 @@ def add_marker():
 
         x, y = Charting.get_cartesean_coords(distance, depth, heading)
 
-        new_marker = Marker(bearing=heading, distance=distance, depth=depth, name=marker_name, marker_type_name=marker_type_name, color=color, x=x, y=y)
-        current_user.map_data.append(new_marker)
-        current_user.save(cascade=True)
+        existing_marker = UserDataController.find_existing_marker_with_name(current_user, marker_name)
+        if existing_marker:
+            old_type_name = existing_marker.marker_type_name
+            UserDataController.update_marker_type(
+                current_user, old_type_name, marker_type_name, color)
+        else:
+            new_marker = Marker(
+                bearing=heading, distance=distance, depth=depth, x=x, y=y,
+                name=marker_name, marker_type_name=marker_type_name, color=color)
+            current_user.map_data.append(new_marker)
+            current_user.save(cascade=True)
```

Comment:
This function is too long. The db code should move to the controller, which can then be unit tested.

Reply:
Fixed